Flatcar SDK: add experimental prefix builds

t-lo opened the pull request:
This change adds experimental prefix builds to the Flatcar SDK.

Prefix builds use a custom sys prefix path and emerge all binaries and runtime dependencies into that prefix.

This path can then e.g. be shipped as a portable sysext since it includes all dependencies, and has libraries at a custom path so these do not conflict with libraries on target systems.

Prefix uses a staging environment (path) featuring a full-blown development environment, and a "final" environment for installing. Staging and final need to be created using setup_prefix first, which will also create an emerge wrapper to emerge ebuilds into staging and subsequently final. The root fs in final may then e.g. be used to create a distro independent, portable sysext.

See the readme included with this PR, at https://github.com/flatcar/scripts/blob/t-lo/experimental-prefix-support/PREFIX.md, for more information.

This PR is a vetted, rebased,  and cleaned up version of https://github.com/flatcar/scripts/pull/1151 . In the previous PR @pothos had helpful and interesting feedback regarding library linker paths - happy to continue this discussion here! 

That PR in turn was a rebased version of the original Flatcar prefix work by @chewi , @jepio , and @HappyTobi here: https://github.com/chewi/flatcar-distro/tree/hackathon-prefix-builds .

## How to use

Prefix requires cross-boss to be available locally.
Tested with the nightly SDK version `3740.0.0+nightly-20230927-2100` (the nightly after that failed so there's no newer SDK available).
The latest development is in the `t-lo/experimental-prefix-support` branch.
```bash
git clone https://github.com/chewi/flatcar-distro.git
cd flatcar-distro
git checkout t-lo/experimental-prefix-support

git clone https://github.com/chewi/cross-boss

./run_sdk_container -t -V 3740.0.0+nightly-20230927-2100
```

In the SDK container, initialise a prefix, then build (for example) python into it:
```bash
# Create prefix staging and final; bootstrap staging env
# Syntax is "setup_prefix <name> <prefix-path>
./setup_prefix prefixtest /usr/local/
emerge-prefix-prefixtest-amd64-usr python
```

Prefixes are architecture aware; the same prefix name can be used for multiple archs
```
./setup_prefix --board=arm64-usr prefixtest /usr/local/
emerge-prefix-prefixtest-arm64-usr python
```

"Final" environments will be in `__prefix__/<board>/prefixtest` after the builds conclude. `__prefix__/<board>/prefixtest/root` contain the root FS and can be used to create sysexts.

## Testing done

Built the above (see "How to use") for amd64 and arm64, created a python sysext, tested on Flatcar x86_64 and arm64 QEmu instances.

Also, added upstream Gentoo as a 4th overlay to a prefix and was able to seamlessly emerge Gentoo packages. Tested a "midnight commander" sysext on both amd64 and arm64 qemu instances.

## Advanced usage

Developers can actually add Gentoo as a custom overlay to their prefix once it was set up.
* Assuming a prefix `my-testprefix` with path `/usr/local/pfx`, simply:
  * create `/build/prefix-amd64-usr/my-testprefix/root/usr/local/pfx/etc/portage/repos.conf/gentoo.conf` and `/home/sdk/trunk/src/scripts/__prefix__/amd64-usr/my-testprefix/root/usr/local/pfx/etc/portage/repos.conf/gentoo.conf`
     with
     ```
     [gentoo]
     location = /mnt/host/source/src/third_party/gentoo
     priority = -1000
     ```
  * clone Gentoo
     ```
     git clone --depth=1 https://github.com/gentoo/gentoo.git sdk_container/src/third_party/gentoo
     ```
* Now you can use `emerge-prefix-my-testprefix-amd64-usr app-misc/mc` to emerge midnight commander from Gentoo.


- [x] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)
- n/a Inspected CI output for image differences: `/boot` and `/usr` size, packages, list files for any missing binaries, kernel modules, config files, kernel modules, etc.

dongsupark commented:
That is the same issue as the recent CI failures of cgroup v1 tests with Kernel 6.1.54.
@jepio reported [upstream](https://lore.kernel.org/stable/20230920081101.GA12096@linuxonhyperv3.guj3yctzbm1etfxqx2vob5hsef.xx.internal.cloudapp.net/T/#u).

t-lo commented:
> Cgroupsv1 and Cilium tests failed: https://github.com/flatcar/scripts/actions/runs/6246234944?pr=1158 (and the test results markdown is too large to be added as a comment here, lol) - but this seems unrelated to this PR?
> 
> All other tests pass.

As per Dongsu's comment the cgroupsv1 issue is unrelated, and all other tests pass. The PR should be ready for review.

t-lo commented:
Rebased on latest main to resolve a conflict with glibc ebuilds.

t-lo commented:
Rebased on the latest main.
This allows folks to build prefixes using the nightly SDK again - wasn't possible yesterday because the weekend nightly SDK build failed.

t-lo commented:
Rebased on latest main to ingest the kernel update that fixes the cgroups v1 issue.
Kicking off an new CI build.

t-lo commented on build_library/prefix_util.sh:
```diff
@@ -0,0 +1,217 @@
+      ;;
+    arm64-usr)
+      PREFIX_CHOST="aarch64-cros-linux-gnu"
+      PREFIX_KEYWORDS="arm64 -~arm64 -multilib"
```
I believe multilib is implicitly disabled via the prefix profile in amd64, while arm64 is not. I'll explicitly disable it here for amd64 too.

t-lo replied:
Removed `-multilib`.

t-lo commented on build_library/prefix_util.sh:
```diff
@@ -0,0 +1,217 @@
+    echo "Skipping install into final as requested."
+else
+    echo "Installing..."
+    sudo -E EPREFIX="\${EPREFIX}" \\
+            ROOT="\${FINALROOT}" \\
+            PORTAGE_CONFIGROOT="\${FINALROOT}\${EPREFIX}" emerge "\$@"
```
`SYSROOT` would be `FINALROOT` but isn't strictly needed here because we're installing binary packages where `SYSROOT` doesn't matter.

t-lo commented on sdk_container/src/third_party/coreos-overlay/sys-libs/glibc/glibc-2.37-r5.ebuild:
```diff
@@ -447,6 +447,10 @@ setup_flags() {
 	# https://sourceware.org/PR27837
 	filter-ldflags '-Wl,--relax'
 
+	# Flag added for cross-prefix, but causes ldconfig to segfault. Not needed
+	# anyway because glibc already handles this by itself.
+	filter-ldflags '-Wl,--dynamic-linker=*'
+
```
As far as I understood @chewi this fix is pending upstream. If it's not there in the next bump we're doing for glibc we should upstream it.

chewi replied:
It was already upstream, although I didn't apply it as far back as 2.37.

t-lo replied:
Ah, so it's a backport. We'll pick it up when we bump glibc versions again.

t-lo commented on sdk_container/src/third_party/portage-stable/profiles/features/prefix/standalone/profile.bashrc:
```diff
@@ -20,7 +20,7 @@ elif [[ ${CATEGORY}/${PN} == sys-libs/glibc && ${EBUILD_PHASE} == configure ]];
     for f in libio/iopopen.c \
 		 shadow/lckpwdf.c resolv/{netdb,resolv}.h elf/rtld.c \
-		 nis/nss_compat/compat-{grp,initgroups,{,s}pwd}.c \
+		 n[is]s/nss_compat/compat-{grp,initgroups,{,s}pwd}.c \
```
Yes, this should trickle down eventually.

t-lo commented on sdk_container/src/third_party/portage-stable/sys-apps/bubblewrap/bubblewrap-0.8.0.ebuild:
Will do!

t-lo replied:
Done.

chewi commented on sdk_container/src/third_party/prefix-overlay/skel/etc/portage/package.accept_keywords/portage:
```diff
@@ -0,0 +1 @@
+sys-apps/portage ~amd64
```
This shouldn't be necessary now that 3.0.51 is stable. At least 3.0.50 was needed for the transition to Meson, which was primarily motivated by this work.

t-lo replied:
Unfortunately this is necessary for cross-boss' cb-bootstrap to work. It's only used for bootstrapping the "staging" area and explicitly disabled via PREFIX_KEYWORDS when emerging into the prefix.

t-lo replied:
gcc-13 is unstable too, right? But if we switch to gcc-13 then we can remove the above.

krnowak replied:
Stable on amd64, unstable on arm64. We have some accept keywords in coreos-overlay, I suppose that prefix-overlay will need its own.

chewi commented on sdk_container/src/third_party/prefix-overlay/skel/etc/portage/package.mask/coreos:
```diff
@@ -0,0 +1,2 @@
+app-misc/ca-certificates::coreos # coreos ca-certificates doesn't support prefix
+sys-apps/baselayout::coreos # coreos baselayout depends on systemd
```
The packages are fundamentally different to their upstream counterparts, which is a shame. I feel like they don't have to differ quite so much. The coreos baselayout could possibly even be an additional package instead.

t-lo replied:
It depends on the feasibility - but it's sure worth researching. Baselayout in coreos-overlay in particular is very Flatcar specific; if we can manage to use stock Gentoo baselayout then this can be removed here.

krnowak replied:
Using Gentoo baselayout instead of our own stuff would be nice. Maybe it's even possible, who knows. Just no one had a look at it yet. We have a bunch of our own stuff in our baselayout fork, though.

krnowak replied:
But until then, I suppose we would need to do some work on our fork to support eprefix in order to make the prefix overlay come out of experimental status.

t-lo commented on sdk_container/src/third_party/prefix-overlay/app-misc/ca-certificates/ca-certificates-20230311.3.93.ebuild:
coreos-overlay currently does not support prefix.

t-lo commented on sdk_container/src/third_party/prefix-overlay/dev-lang/python/python-2.7.18_p16-r1.ebuild:
The whole of dev-lang/python an import from upstream Gentoo to fix a prefix build issue.
Integrating it with portage-stable is part of the prefix stabilisation process; for now prefix is experimental.

t-lo replied:
Wait, I just checked, this has no differences to the python in portage-stable. I've removed it and re-rested the prefix; still works fine.

t-lo commented on sdk_container/src/third_party/prefix-overlay/eclass/distutils-r1.eclass:
This is an import from upstream Gentoo to fix a prefix build issue.
Integrating it with portage-stable is part of the prefix stabilisation process; for now prefix is experimental.

t-lo commented on sdk_container/src/third_party/prefix-overlay/eclass/python-utils-r1.eclass:
This is an import from upstream Gentoo to fix a prefix build issue.
Integrating it with portage-stable is part of the prefix stabilisation process; for now prefix is experimental.

t-lo replied:
(Also, we carry prefix specific patches on top of upstream)

chewi commented on sdk_container/src/third_party/prefix-overlay/dev-libs/libgpg-error/libgpg-error-1.47.ebuild:
```diff
@@ -0,0 +1,95 @@
+		hprefixify -e "s#${EPREFIX}#${BROOT}#" -w 1 autogen.sh
+	else
+		hprefixify -w 1 autogen.sh
+	fi
```
I have proposed a better fix for this upstream. See gentoo/gentoo#33010.

t-lo replied:
I've updated the ebuild to use the fix you've proposed.

t-lo commented on sdk_container/src/third_party/prefix-overlay/skel/etc/portage/package.mask/cross:
```diff
@@ -0,0 +1 @@
+>=sys-devel/gcc-13 # FIXME: Cannot cross-compile gcc 13+ using older version
```
Hopefully! For now it's required to make bootstrapping work. BTW if we switch to gcc-13 then I believe we can also remove the unstable un-mask from bootstrapping.

t-lo commented on build_library/prefix_util.sh:
```diff
@@ -0,0 +1,217 @@
+FINALROOT="${FINALROOT}"
+CB_ROOT="${CB_ROOT}"
+
+if [ "\${1}" = "--help" ] ; then
```
Gotcha, will do.

t-lo commented:
> I'm not entirely sure if another overlay is really needed. I think that ideally everything we need is in coreos-overlay for our stuff and in portage-stable for stuff straight from Gentoo.

Absolutely agree - for the first (experimental) prefix support in the SDK these are a (temporaty) necessity for 2 reasons:
- we need to pull in a few updates from upstream for prefix bootstrapping to work
- we need to carry a few patches on top of upstream to fix prefix builds.

We aim for prefix-overlay to eventually be merged into our regular overlays. Keeping it separate for now allows us to ingest the prefix feature and ship it with the SDK, which will enable more devs to work on this feature.
 
> Not a fan of having duplicated packages in prefix overlay. I assume this is temporary as they have some fixes for prefix builds that will eventually find their way to either portage-stable or coreos-overlay (I'm looking at app-misc/ca-certificates and sys-apps/baselayout here).

All packages in prefix-overlay need to be justified - if it's just duplicates then we should not carry them.

> Also, is this really Flatcar specific? I mean - it uses Flatcar tooling and build scripts, but looks like a thing you could pull off using Gentoo repo. I suppose it could be made generic to support building stuff either from Gentoo as primary + extra overlays or from portage-stable as primary + coreos-overlay.

This feature is meant to extend the scope of the SDK to also build distro independent packages. It's not Flatcar specific but it uses the portability of our SDK container and the well-defined environment we provide in our SDK releases to make prefix builds approachable and easy for a larger developer audience.

t-lo commented on sdk_container/src/third_party/portage-stable/virtual/tmpfiles/tmpfiles-0-r5.ebuild:
Done.

t-lo commented:
Cleaned up (removed python from prefix; wasn't necessary after all), rebased on latest main, and rebuilt locally.

I've added Gentoo as a custom overlay to my test prefix repos.conf and was able to successfully emerge midnight commander. Tested on qemu flatcar (amd64):

![image](https://github.com/flatcar/scripts/assets/71888/5c96fd21-49b4-40e4-acf1-0e6cc96fced2)

t-lo commented:
**NOTE** that for testing the PR you'll need to use the nightly SDK from 2 nights ago - last night's build failed and there is no SDK. To do this, start the container like this:
```
./run_sdk_container -t -V 3740.0.0+nightly-20230927-2100
```

krnowak reviewed (approved):
I'd say let's get this in. Ideally there would be some automatic thing going on (like a github action), so it would catch breakage quickly before it bitrots too much.

t-lo commented:
> I'd say let's get this in. Ideally there would be some automatic thing going on (like a github action), so it would catch breakage quickly before it bitrots too much.

Agreed @krnowak . I've created https://github.com/flatcar/Flatcar/issues/1195 for tracking and added it to the relevant boards.
